okio-files/src/commonMain/kotlin/okio/FileMetadata.kt

```diff
@@ -0,0 +1,105 @@
+ * path may refer to:
+ *
+ *  * An operating system process that consumes data, produces data, or both. For example, reading
+ *    from the `/dev/urandom` file on Linux returns a unique sequence of pseudorandom bytes to each
+ *    reader.
+ *
+ *  * A stream that connects a pair of programs together. A pipe is a special file that a producing
+ *    program writes to and a consuming program reads from. Both programs operate concurrently. The
+ *    size of a pipe is not well defined: the writer can write as much data as the reader is able to
+ *    read.
+ *
+ *  * A file on a remote filesystem. The performance and availability of remote files may be quite
+ *    different from that of local files!
+ *
+ *  * A symbolic link (symlink) to another path. When attempting to access this path the filesystem
+ *    will follow the link and return data from the target path.
+ *
+ *  * The same content as another path without a symlink. On UNIX filesystems an inode is an
+ *    anonymous handle to a file's content, and multiple paths may target the same inode without any
+ *    other relationship to one another. A consequence of this design is that a directory with three
+ *    1 GiB files may only need 1 GiB on the storage device.
+ *
+ * This class does not attempt to model these rich filesystem features! It exposes a limited view
+ * useful for programs with only basic filesystem needs. Be cautious of the potential consequences
+ * of special files when writing programs that operate on a filesystem.
+ *
+ * File metadata is subject to change, and code that operates on filesystems should defend against
+ * changes to the file that occur between reading metadata and subsequent operations.
+ */
+class FileMetadata(
+  /** True if this file is a container of bytes. If this is true, then [size] is non-null. */
+  val isRegularFile: Boolean,
+
+  /** True if the path refers to a directory that contains 0 or more child paths. */
+  val isDirectory: Boolean,
+
+  /**
+   * Returns the number of bytes readable from this file. The amount of storage resources consumed
+   * by this file may be larger (due to block size overhead, redundant copies for RAID, etc.), or
+   * smaller (due to filesystem compression, shared inodes, etc).
+   */
+  val size: Long?,
+
+  /**
+   * Returns the system time of the host computer when this file was created, if the host filesystem
+   * supports this feature. This is typically available on Windows NTFS filesystems and not
+   * available on UNIX or Windows FAT filesystems.
+   */
+  val createdAtMillis: Long?,
+
+  /**
+   * Returns the system time of the host computer when this file was most recently written.
+   *
+   * Note that the accuracy of the returned time may be much more coarse than its precision. In
+   * particular, this value is is expressed with millisecond precision but may be accessed at
+   * second- or day-accuracy only.
+   */
+  val lastModifiedAtMillis: Long?,
```

Comment:
Same as with `size`, maybe worth being more explicit about the cases when this is expected to be null. Also applies to `lastAccessedAtMillis`.

Reply:
I’d love to be more explicit, but the guarantees offered by the underlying systems are extremely weak and filesystem dependent. One trivial example: a file on a readonly media like a CD-ROM doesn’t have a mechanism to track last access time.

https://docs.microsoft.com/en-us/windows/win32/sysinfo/file-times

> Not all file systems can record creation and last access times, and not all file systems record them in the same manner. For example, the resolution of create time on FAT is 10 milliseconds, while write time has a resolution of 2 seconds and access time has a resolution of 1 day, so it is really the access date. The NTFS file system delays updates to the last access time for a file by up to 1 hour after the last access.

Reply:
Interesting! But makes sense